code/modules/xenoarcheaology/effect_master.dm

```diff
@@ -0,0 +1,383 @@
+			active_effects |= my_effect
+
+	return active_effects
+
+/datum/component/artifact_master/proc/add_effect()
+	var/effect_type = input(usr, "What type do you want?", "Effect Type") as null|anything in typesof(/datum/artifact_effect) - /datum/artifact_effect
+	if(effect_type)
+		var/datum/artifact_effect/my_effect = new effect_type(src)
+		if(istype(holder, my_effect.req_type))
+			my_effects += my_effect
+
+		else
+			to_chat(usr, "This effect can not be applied to this atom type.")
+			qdel(my_effect)
+
+/datum/component/artifact_master/proc/remove_effect()
+	var/to_remove_effect = input(usr, "What effect do you want to remove?", "Remove Effect") as null|anything in my_effects
+
+	if(to_remove_effect)
+		var/datum/artifact_effect/AE = to_remove_effect
+		my_effects.Remove(to_remove_effect)
+		qdel(AE)
+
+/datum/component/artifact_master/Destroy()
+	holder = null
+	for(var/datum/artifact_effect/AE in my_effects)
+		AE.master = null
+		qdel(AE)
+
+	STOP_PROCESSING(SSobj,src)
+
+	. = ..()
+
+/datum/component/artifact_master/proc/do_setup()
+	if(LAZYLEN(make_effects))
+		for(var/path in make_effects)
+			var/datum/artifact_effect/new_effect = new path(src)
+			if(istype(holder, new_effect.req_type))
+				my_effects += new_effect
+
+	else
+		generate_effects()
+
+	DoRegistry()
+
+/datum/component/artifact_master/proc/generate_effects()
+	while(effect_generation_chance > 0)
+		var/chosen_path = pick(subtypesof(/datum/artifact_effect))
+		if(effect_generation_chance >= 100)	// If we're above 100 percent, just cut a flat amount and add an effect.
+			var/datum/artifact_effect/AE = new chosen_path(src)
+			if(istype(holder, AE.req_type))
+				my_effects += AE
+				effect_generation_chance -= 30
+			else
+				AE.master = src
+				qdel(AE)
+			continue
+
+		effect_generation_chance /= 2
+
+		if(prob(effect_generation_chance))	// Otherwise, add effects as normal, with decreasing probability.
+			my_effects += new chosen_path(src)
+
+/datum/component/artifact_master/proc/get_holder()	// Returns the holder.
+	return holder
+
+/datum/component/artifact_master/proc/get_primary()
+	if(LAZYLEN(my_effects))
+		return my_effects[1]
+	return FALSE
+
+/*
+ * Trigger code.
+ */
+
+/datum/component/artifact_master/proc/on_exact()
+	var/severity = args[2]
+	var/triggered = FALSE
+	for(var/datum/artifact_effect/my_effect in my_effects)
+		switch(severity)
+			if(1.0)
+				if(my_effect.trigger == TRIGGER_FORCE || my_effect.trigger == TRIGGER_HEAT ||  my_effect.trigger == TRIGGER_ENERGY)
+					my_effect.ToggleActivate()
+					triggered = TRUE
+			if(2.0)
+				if(my_effect.trigger == TRIGGER_FORCE || my_effect.trigger == TRIGGER_HEAT)
+					my_effect.ToggleActivate()
+					triggered = TRUE
+			if(3.0)
+				if (my_effect.trigger == TRIGGER_FORCE)
+					my_effect.ToggleActivate()
+					triggered = TRUE
+
+	if(triggered)
+		return COMPONENT_IGNORE_EXPLOSION
+
+	return
+
+/datum/component/artifact_master/proc/on_bullet()
+	var/obj/item/projectile/P = args[2]
```

Comment:
does the callback usage prevent you from outright naming the arguments as parameters?

Reply:
Due to the system, I'd had trouble getting it to work right without manually specifying from the args list.